### ISSUE TITLE
Do you pass the Carry Flag result of the int 13h call to your caller anywhere?

### ISSUE THREAD

**Issue by ecm-pushbx** [MobyGamer/softhddi#1]
Looking at this part: https://github.com/MobyGamer/softhddi/blob/8f2a6f2d81b32911f7a279925abebfc11bdde41f/SOFTHDDI.ASM#L869

It seems like you do not use the CF result from the int 13h call. Am I missing something?

**Comment by ecm-pushbx** [comment]
Technically the `pushf` before and the int 13h call itself should also get the *incoming* CF from your caller. But that's more subtle of a point than the outgoing CF.

**Comment by MobyGamer** [comment]
Good call (pun intended), and something I'm embarrased I missed.  CF is set on return from most functions if there is an error, so that needs to be passed on.  I'll submit a fix shortly.

**Comment by ecm-pushbx** [comment]
You didn't understand the extent of this error, and your two fixes do not actually fix the main problem. The `iret` instruction pops flags from the stack. In order to return a CF result, you must not use only `iret` without modifying the `fl` word on the `iret` stack frame. Either do the latter, or use `retf 2` to return instead of `iret`.

Fix would look like:

```
ISP_HEADER 13h
        pushf                           ;preserve flags on entry
        push    bx                      ;need an index register
        IF OMITFLOPPY
        cmp     dl,80h                  ;is the target for this call a HDD?
        jb      noindicators            ;if not, skip indicators
        ENDIF
        cmp     ah,0Fh                  ;is this outside our understood calls?
        ja      noindicators            ;if so, just do original int13
        xor     bx,bx
        mov     bl,ah
        cmp     byte ptr [bx+dsvctbl],1 ;does this call do physical activity?
        jne     noindicators            ;if not, don't do any indicators

doindicators:
        pop     bx                      ;restore changed bx
        call    diskIntEnter            ;start indicators

        popf ; return CF to what we received as input
        ;Simulate interrupt call to original INT 13h
        pushf ; build an interrupt-like stack frame
        call    ORIG_INT13h             ;pass to original disk interrupt
        pushf                           ;save int13 result flags
        call    diskIntLeave            ;stop indicators
        popf                            ;restore int13 result flags
push bp
mov bp, sp
push ax
lahf
; bp + 0 = saved bp
; bp + 2 = ip
; bp + 4 = cs
; bp + 6 = fl
mov byte [bp + 6], ah
pop ax
pop bp
iret

noindicators:
        pop     bx                      ;restore potentially-changed bx
        popf                            ;restore flags from entry
        jmp     ORIG_INT13h
```

Stack frame modification copied from what I suggested in https://github.com/640-KB/GLaBIOS/issues/20

Alternative that only passes CF and not ZF etc in https://hg.pushbx.org/ecm/seekext/file/bbfcfa0d1c0b/resident.asm#l254

**Comment by MobyGamer** [comment]
You're right, the flags are not being patched into the stack.  I've re-opened the issue and removed all releases.

I want to note that this bug won't actually result in data loss, as int 13h calls only use cf to report error, and if you have an error of this nature in DOS, the system almost always hangs waiting for I/O (which is why I didn't find it in testing, and I defragged a hard disk using 1.0 of the code).  That said, I've removed all releases until this is fixed.
